File: examples/e2e-testing/rest/java/E2ETesting.java
// Runs an end-to-end test against a remote Browserless browser using BQL.
//
// Run: javac E2ETesting.java && java E2ETesting

import java.net.URI;
import java.net.http.*;

public class E2ETesting {
    public static void main(String[] args) throws Exception {
        String token = "YOUR_API_TOKEN_HERE";
        String endpoint = "https://production-sfo.browserless.io/bql?token=" + token;

        String query = "mutation E2ETest {"
            + " goto(url: \\\"https://automationexercise.com\\\", waitUntil: networkIdle) { status }"
            + " title { title }"
            + " verify: text(selector: \\\"h2\\\") { text }"
            + " }";

        String payload = "{\"query\": \"" + query + "\", \"variables\": {}, \"operationName\": \"E2ETest\"}";

        HttpClient client = HttpClient.newHttpClient();
        HttpRequest request = HttpRequest.newBuilder()
            .uri(URI.create(endpoint))
            .header("Content-Type", "application/json")
            .POST(HttpRequest.BodyPublishers.ofString(payload))
            .build();

        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
        String body = response.body();
        System.out.println(body);

        if (!body.contains("\"status\":200")) {
            throw new AssertionError("Expected status 200");
        }
        System.out.println("E2E test passed.");
    }
}

File: examples/e2e-testing/rest/csharp/E2ETesting.cs
// Runs an end-to-end test against a remote Browserless browser using BQL.
//
// Run: dotnet run

using System.Net.Http;
using System.Text;
using System.Text.Json;

string token = "YOUR_API_TOKEN_HERE";
string endpoint = $"https://production-sfo.browserless.io/bql?token={token}";

var payload = new
{
    query = @"mutation E2ETest {
      goto(url: ""https://automationexercise.com"", waitUntil: networkIdle) { status }
      title { title }
      verify: text(selector: ""h2"") { text }
    }",
    variables = new { },
    operationName = "E2ETest",
};

using (HttpClient httpClient = new HttpClient())
{
    var content = new StringContent(
        JsonSerializer.Serialize(payload), Encoding.UTF8, "application/json");
    var response = await httpClient.PostAsync(endpoint, content);
    string body = await response.Content.ReadAsStringAsync();

    var doc = JsonDocument.Parse(body);
    int status = doc.RootElement.GetProperty("data").GetProperty("goto").GetProperty("status").GetInt32();
    if (status != 200) throw new Exception($"Expected status 200, got {status}");

    string title = doc.RootElement.GetProperty("data").GetProperty("title").GetProperty("title").GetString() ?? "";
    if (!title.Contains("Automation")) throw new Exception($"Title check failed: {title}");

    Console.WriteLine("E2E test passed.");
    Console.WriteLine($"Page title: {title}");
}
